## Project 1 Data Cleaning

Goals:
- import and examine data
- make structural changes to the data necessary
- merge red and white wine data files
- save as new csv

In [1]:
#import libraries
import pandas as pd #work with dataframes
import numpy as np #perform mathematical manipulations
import requests #Retrieve data
import io #interpret data encoding

import timeit #time notebook
start_time = timeit.default_timer()


In [2]:
red="https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv"
redwine=requests.get(red).content
reddf=pd.read_csv(io.StringIO(redwine.decode('utf-8')))

white="https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv"
whitewine=requests.get(white).content
whitedf=pd.read_csv(io.StringIO(whitewine.decode('utf-8'))) #solution to error with original method (pd.read_csv(whitewine)) found using https://stackoverflow.com/questions/32400867/pandas-read-csv-from-url

In [3]:
reddf.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
0   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                     
1   7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                                                     
2  7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                                                     
3  11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                                                     
4   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5

In [4]:
whitedf.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
0   7;0.27;0.36;20.7;0.045;45;170;1.001;3;0.45;8.8;6                                                                                                                     
1  6.3;0.3;0.34;1.6;0.049;14;132;0.994;3.3;0.49;9...                                                                                                                     
2  8.1;0.28;0.4;6.9;0.05;30;97;0.9951;3.26;0.44;1...                                                                                                                     
3  7.2;0.23;0.32;8.5;0.058;47;186;0.9956;3.19;0.4...                                                                                                                     
4  7.2;0.23;0.32;8.5;0.058;47;186;0.9956;3.19;0.4...

In [5]:
# I need to expand the dfs into multiple columns with each feature represented as 1 column
reddf[["fixed acidity","volatile acidity","citric acid","residual sugar","chlorides","free sulfur dioxide","total sulfur dioxide","density","pH","sulphates","alcohol","quality"]] = reddf['fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"'].str.split(';', expand = True)
whitedf[["fixed acidity","volatile acidity","citric acid","residual sugar","chlorides","free sulfur dioxide","total sulfur dioxide","density","pH","sulphates","alcohol","quality"]] = whitedf['fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"'].str.split(';', expand = True)

In [6]:
reddf.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"  \
0   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                        
1   7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                                                        
2  7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                                                        
3  11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                                                        
4   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                        

  fixed acidity volatile acidity citric acid residual sugar chlorides  \
0           7.4              0.7           0            1.9     0.076   
1           7.8             0.88           0            2.6     0.098   
2           7.8             0.76        0.04            2.3     0.092   
3          11.2             0.28        0.56            1.9     0.075   
4           7.4              0.7           0            1.9     0.076   

  free sulfur dioxide total sulfur dioxide density    pH sulphates alcohol  \
0                  11                   34  0.9978  3.51      0.56     9.4   
1                  25                   67  0.9968   3.2      0.68     9.8   
2                  15                   54   0.997  3.26      0.65     9.8   
3                  17                   60   0.998  3.16      0.58     9.8   
4                  11                   34  0.9978  3.51      0.56     9.4   

  quality  
0       5  
1       5  
2       5  
3       6  
4       5

In [7]:
whitedf.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"  \
0   7;0.27;0.36;20.7;0.045;45;170;1.001;3;0.45;8.8;6                                                                                                                        
1  6.3;0.3;0.34;1.6;0.049;14;132;0.994;3.3;0.49;9...                                                                                                                        
2  8.1;0.28;0.4;6.9;0.05;30;97;0.9951;3.26;0.44;1...                                                                                                                        
3  7.2;0.23;0.32;8.5;0.058;47;186;0.9956;3.19;0.4...                                                                                                                        
4  7.2;0.23;0.32;8.5;0.058;47;186;0.9956;3.19;0.4...                                                                                                                        

  fixed acidity volatile acidity citric acid residual sugar chlorides  \
0             7             0.27        0.36           20.7     0.045   
1           6.3              0.3        0.34            1.6     0.049   
2           8.1             0.28         0.4            6.9      0.05   
3           7.2             0.23        0.32            8.5     0.058   
4           7.2             0.23        0.32            8.5     0.058   

  free sulfur dioxide total sulfur dioxide density    pH sulphates alcohol  \
0                  45                  170   1.001     3      0.45     8.8   
1                  14                  132   0.994   3.3      0.49     9.5   
2                  30                   97  0.9951  3.26      0.44    10.1   
3                  47                  186  0.9956  3.19       0.4     9.9   
4                  47                  186  0.9956  3.19       0.4     9.9   

  quality  
0       6  
1       6  
2       6  
3       6  
4       6

In [8]:
#add a variable indicating wine type:
whitedf['type'] = [0 for x in range(len(whitedf))]
reddf['type'] = [1 for x in range(len(reddf))]

In [9]:
reddf.head()

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"  \
0   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                        
1   7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                                                        
2  7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                                                        
3  11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                                                        
4   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                        

  fixed acidity volatile acidity citric acid residual sugar chlorides  \
0           7.4              0.7           0            1.9     0.076   
1           7.8             0.88           0            2.6     0.098   
2           7.8             0.76        0.04            2.3     0.092   
3          11.2             0.28        0.56            1.9     0.075   
4           7.4              0.7           0            1.9     0.076   

  free sulfur dioxide total sulfur dioxide density    pH sulphates alcohol  \
0                  11                   34  0.9978  3.51      0.56     9.4   
1                  25                   67  0.9968   3.2      0.68     9.8   
2                  15                   54   0.997  3.26      0.65     9.8   
3                  17                   60   0.998  3.16      0.58     9.8   
4                  11                   34  0.9978  3.51      0.56     9.4   

  quality  type  
0       5     1  
1       5     1  
2       5     1  
3       6     1  
4       5     1

In [10]:
#Drop the original columns and merge the data into one wine dataframe which will be used to create a model predicting wine quality
reddf.drop('fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"', axis=1, inplace = True)
whitedf.drop('fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"', axis=1, inplace = True)


In [11]:
reddf.head()

fixed acidity volatile acidity citric acid residual sugar chlorides  \
0           7.4              0.7           0            1.9     0.076   
1           7.8             0.88           0            2.6     0.098   
2           7.8             0.76        0.04            2.3     0.092   
3          11.2             0.28        0.56            1.9     0.075   
4           7.4              0.7           0            1.9     0.076   

  free sulfur dioxide total sulfur dioxide density    pH sulphates alcohol  \
0                  11                   34  0.9978  3.51      0.56     9.4   
1                  25                   67  0.9968   3.2      0.68     9.8   
2                  15                   54   0.997  3.26      0.65     9.8   
3                  17                   60   0.998  3.16      0.58     9.8   
4                  11                   34  0.9978  3.51      0.56     9.4   

  quality  type  
0       5     1  
1       5     1  
2       5     1  
3       6     1  
4       5     1

In [12]:
#join datasets
winesdf = reddf.append(whitedf)

In [13]:
winesdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6497 entries, 0 to 4897
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   fixed acidity         6497 non-null   object
 1   volatile acidity      6497 non-null   object
 2   citric acid           6497 non-null   object
 3   residual sugar        6497 non-null   object
 4   chlorides             6497 non-null   object
 5   free sulfur dioxide   6497 non-null   object
 6   total sulfur dioxide  6497 non-null   object
 7   density               6497 non-null   object
 8   pH                    6497 non-null   object
 9   sulphates             6497 non-null   object
 10  alcohol               6497 non-null   object
 11  quality               6497 non-null   object
 12  type                  6497 non-null   int64 
dtypes: int64(1), object(12)
memory usage: 710.6+ KB


In [14]:
print(reddf.shape, whitedf.shape, winesdf.shape)

(1599, 13) (4898, 13) (6497, 13)


I want to classify which wines are good, okay and bad using this dataset, so I will use the quality scale to divide these into 3 labels for simplicity.

In [15]:
winesdf['quality'] = winesdf['quality'].astype(int)
categories = []
for i in winesdf['quality']:
    if i <=4:
        categories.append('bad wine')
    elif i >=7:
        categories.append('good wine')
    else:
        categories.append('just okay wine')
        
winesdf['categories'] = categories

In [16]:
winesdf.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'type', 'categories'],
      dtype='object')

In [17]:
wines = winesdf[['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'type', 'quality']].astype(float)
wines.corr()

fixed acidity  volatile acidity  citric acid  \
fixed acidity              1.000000          0.219008     0.324436   
volatile acidity           0.219008          1.000000    -0.377981   
citric acid                0.324436         -0.377981     1.000000   
residual sugar            -0.111981         -0.196011     0.142451   
chlorides                  0.298195          0.377124     0.038998   
free sulfur dioxide       -0.282735         -0.352557     0.133126   
total sulfur dioxide      -0.329054         -0.414476     0.195242   
density                    0.458910          0.271296     0.096154   
pH                        -0.252700          0.261454    -0.329808   
sulphates                  0.299568          0.225984     0.056197   
alcohol                   -0.095452         -0.037640    -0.010493   
type                       0.486740          0.653036    -0.187397   
quality                   -0.076743         -0.265699     0.085532   

                      residual sugar  chlorides  free sulfur dioxide  \
fixed acidity              -0.111981   0.298195            -0.282735   
volatile acidity           -0.196011   0.377124            -0.352557   
citric acid                 0.142451   0.038998             0.133126   
residual sugar              1.000000  -0.128940             0.402871   
chlorides                  -0.128940   1.000000            -0.195045   
free sulfur dioxide         0.402871  -0.195045             1.000000   
total sulfur dioxide        0.495482  -0.279630             0.720934   
density                     0.552517   0.362615             0.025717   
pH                         -0.267320   0.044708            -0.145854   
sulphates                  -0.185927   0.395593            -0.188457   
alcohol                    -0.359415  -0.256916            -0.179838   
type                       -0.348821   0.512678            -0.471644   
quality                    -0.036980  -0.200666             0.055463   

                      total sulfur dioxide   density        pH  sulphates  \
fixed acidity                    -0.329054  0.458910 -0.252700   0.299568   
volatile acidity                 -0.414476  0.271296  0.261454   0.225984   
citric acid                       0.195242  0.096154 -0.329808   0.056197   
residual sugar                    0.495482  0.552517 -0.267320  -0.185927   
chlorides                        -0.279630  0.362615  0.044708   0.395593   
free sulfur dioxide               0.720934  0.025717 -0.145854  -0.188457   
total sulfur dioxide              1.000000  0.032395 -0.238413  -0.275727   
density                           0.032395  1.000000  0.011686   0.259478   
pH                               -0.238413  0.011686  1.000000   0.192123   
sulphates                        -0.275727  0.259478  0.192123   1.000000   
alcohol                          -0.265740 -0.686745  0.121248  -0.003029   
type                             -0.700357  0.390645  0.329129   0.487218   
quality                          -0.041385 -0.305858  0.019506   0.038485   

                       alcohol      type   quality  
fixed acidity        -0.095452  0.486740 -0.076743  
volatile acidity     -0.037640  0.653036 -0.265699  
citric acid          -0.010493 -0.187397  0.085532  
residual sugar       -0.359415 -0.348821 -0.036980  
chlorides            -0.256916  0.512678 -0.200666  
free sulfur dioxide  -0.179838 -0.471644  0.055463  
total sulfur dioxide -0.265740 -0.700357 -0.041385  
density              -0.686745  0.390645 -0.305858  
pH                    0.121248  0.329129  0.019506  
sulphates            -0.003029  0.487218  0.038485  
alcohol               1.000000 -0.032970  0.444319  
type                 -0.032970  1.000000 -0.119323  
quality               0.444319 -0.119323  1.000000

Based on the correlation matrix above, quality appears to be most related to volitile acidity, chlorides, density, alcohol, and potentially type. I will keep this in mind during the exploratory analysis phase.

My final data set is 6,497 rows and 13 columns. 12 columns will be used as predictor features for the wine quality in the model. My last step for cleaning will be to examine the data for nulls, and make any necessary corrections.

In [18]:
winesdf.isnull().sum() #counts the nulls for each feature

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
type                    0
categories              0
dtype: int64

Excellent! No nulls in the data. I will therefore save the final dataset and proceed to data exploration

In [19]:
winesdf.drop('quality', axis =1, inplace = True)
winesdf.to_csv('./WinesData.csv')

In [20]:
elapsed = timeit.default_timer() - start_time
print(str(elapsed) + " seconds")

1.6461009 seconds
